flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/util/SubplanReuseUtil.scala

```diff
@@ -0,0 +1,217 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.table.plan.util
+
+import org.apache.flink.table.api.{PlannerConfigOptions, TableConfig, TableException}
+import org.apache.flink.table.plan.nodes.calcite.Sink
+import org.apache.flink.table.plan.nodes.logical.FlinkLogicalTableSourceScan
+import org.apache.flink.table.plan.nodes.physical.PhysicalTableSourceScan
+
+import com.google.common.collect.{Maps, Sets}
+import org.apache.calcite.rel.core.{Exchange, TableFunctionScan}
+import org.apache.calcite.rel.{RelNode, RelVisitor}
+
+import java.util
+
+import scala.collection.JavaConversions._
+
+object SubplanReuseUtil {
+
+  def reuseSubplan(rels: Seq[RelNode], tableConfig: TableConfig): Seq[RelNode] = {
```

Comment:
Isn't this the major logic of sub plan reusing? I don't think these kind of logic should stay in a utility class.

Reply:
yes. I will rename SubplanReuseUtil to SubplanReuser, and move it to `org.apache.flink.table.plan.reuse`.